README.md

# api-node-test

This is built with Nodejs.

# Goal 

Create an API with token authentification.

# Install

Install with Npm.

```
npm install
```

# Run 

 ```
npm server.js
```

# Release notes

See CHANGELOG.md for a summary of changes.

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project loosely follows [Semantic Versioning](http://semver.org/),
but is still in early stages of development.

## [0.0.1] - 2016-1204

* The app is now named `api-node-test`


server.js
// =======================
// get the packages we need ============
// =======================
var express     = require('express');
var app         = express();
var fileSystem  = require('fs');
var bodyParser  = require('body-parser');
var expressValidator = require('express-validator');
var morgan      = require('morgan');
var mongoose    = require('mongoose');

var jwt    = require('jsonwebtoken'); // used to create, sign, and verify tokens
var config = require('./config'); // get our config file
var UserMongo  = require('./app/models/user.mongo');

// =======================
// configuration =========
// =======================
var port = process.env.PORT || 8080; // used to create, sign, and verify tokens
mongoose.connect(process.env.MONGODB_URI || config.database); // connect to database

app.set('superSecretAdmin', config.secretAdmin); // secret variable
app.set('superSecretUser', config.secretUser); // secret variable

// use body parser so we can get info from POST and/or URL parameters
app.use(bodyParser.urlencoded({ extended: false }));
app.use(bodyParser.json());
app.use(expressValidator());

// use morgan to log requests to the console
app.use(morgan('dev'));

require('./app/routes/api.js')(app, express, jwt) ;
require('./app/routes/admin.js')(app, express, jwt) ;
// apply the routes to our application with the prefix /api


app.get('/', function(req, res) {
    res.send('Hello! The API is at http://localhost:' + port + '/');
});

app.get('/setup', function(req, res) {

  // create a sample user
  var nick = new UserMongo({ 
    name: 'jick', 
    password: 'password',
    admin: true 
  });

  // save the sample user
  nick.save(function(err) {
    if (err) throw err;

    console.log('User saved successfully');
    res.json({ success: true });
  });
});

app.get('/setupUser', function(req, res) {

  // create a sample user
  var nick = new UserMongo({ 
    name: 'jean', 
    password: 'password',
    admin: false 
  });

  // save the sample user
  nick.save(function(err) {
    if (err) throw err;

    console.log('User saved successfully');
    res.json({ success: true });
  });
});
// API ROUTES -------------------
// we'll get to these in a second

// =======================
// start the server ======
// =======================
app.listen(port);
console.log('Magic happens at http://localhost:' + port);

app/routes/api.js
"use strict";

//module.exports = function(io, app, Global, db_produits, db_backup, fs) {
module.exports = function(app, express, jwt ) {

// get an instance of the router for api routes
var apiRoutes = express.Router(); 


var UserMongo  = require('./../models/user.mongo');
// =======================
// routes ================
// =======================
// basic route

// route to authenticate a user (POST http://localhost:8080/api/authenticate)

apiRoutes.post('/authenticate', function(req, res) {

  // find the user
  UserMongo.findOne({
    name: req.body.name
  }, function(err, user) {

    if (err) throw err;

    if (!user) {
      res.json({ success: false, message: 'Authentication failed. User not found.' });
    } else if (user) {

      // check if password matches
      if (user.password != req.body.password) {
        res.json({ success: false, message: 'Authentication failed. Wrong password.' });
      } else {

        // if user is found and password is right
        // create a token

        var token = jwt.sign(user, app.get('superSecretUser'), {
          expiresIn: 1440 // expires in 24 hours
        });

        // return the information including token as JSON
        res.json({
          success: true,
          message: 'Enjoy your token!',
          token: token
        });
      }   

    }

  });
});



// route middleware to verify a token
apiRoutes.use(function(req, res, next) {

  // check header or url parameters or post parameters for token
  var token = req.body.token || req.query.token || req.headers['x-access-token'];

  // decode token
  if (token) {

    // verifies secret and checks exp 
    // simple user  
    jwt.verify(token, app.get('superSecretUser'), function(err, decoded) {      
      if (err) {
        // admin user
        jwt.verify(token, app.get('superSecretAdmin'), function(err, decoded) {      
        if (err) {
          return res.status(403).json({ success: false, message: 'Failed to authenticate token.' });    
        } else {
          // if everything is good, save to request for use in other routes
          req.decoded = decoded;    
          next();
        }
      }); 
      } else {
        // if everything is good, save to request for use in other routes
        req.decoded = decoded;    
        next();
      }
    });

  } else {

    // if there is no token
    // return an error
    return res.status(403).send({ 
        success: false, 
        message: 'No token provided.' 
    });
    
  }
});

apiRoutes.get('/', function(req, res) {
    res.send('Hello! You can access API endpoint!');
});

// apply the routes to our application with the prefix /api
app.use('/api', apiRoutes);

};

app/tools/validator.js
"use strict";

// Functions for format validation
module.exports = {

    // validate authenticate input format 
    validateAuthentificate : function(app, req, res) {
        
        req.checkBody( "name", "Required").notEmpty();

        req.checkBody( "password", "Required").notEmpty();

        var errors = req.validationErrors();
        if (errors) {
            res.status(400).send(errors);
            return;
        }
    },

    // validate add user input format
    validateAddUser : function(app, req, res) {
        
        req.checkBody( "name", "Required").notEmpty();

        req.checkBody( "password", "Required").notEmpty();

        req.checkBody( "admin", "Required").notEmpty();
        req.checkBody( "admin", "Should be boolean").isBoolean();

        var errors = req.validationErrors();
        if (errors) {
            res.status(400).send(errors);
            return;
        }
    },
        // validate add user input format
    validateDeleteUser : function(app, req, res) {
        
        req.checkBody( "name", "Required").notEmpty();

        var errors = req.validationErrors();
        if (errors) {
            res.status(400).send(errors);
            return;
        }
    }
};

app/routes/admin.js
"use strict";

//module.exports = function(io, app, Global, db_produits, db_backup, fs) {
module.exports = function(app, express, jwt ) {
// get an instance of the router for api routes
var adminRoutes = express.Router(); 
//var validatorTool = require('../tools/validator.js');

var UserMongo  = require('./../models/user.mongo');
// =======================
// routes ================
// =======================
// basic route

// route to authenticate a user (POST http://localhost:8080/api/authenticate)

adminRoutes.post('/authenticate', function(req, res) {

  // find the user
  UserMongo.findOne({
    name: req.body.name
  }, function(err, user) {

    if (err) throw err;

    if (!user) {
      res.json({ success: false, message: 'Authentication failed. User not found.' });
    } else if (user) {

      // check if password matches
      if (user.password != req.body.password) {
        res.json({ success: false, message: 'Authentication failed. Wrong password.' });
      } else {

        if (!user.admin) {
        res.json({ success: false, message: 'Authentication failed. You are not admin' });
        } else {
          // if user is found and password is right
          // create a token

          var token = jwt.sign(user, app.get('superSecretAdmin'), {
            expiresIn: 1440 // expires in 24 hours
          });

          // return the information including token as JSON
          res.json({
            success: true,
            message: 'Enjoy your token!',
            token: token
          });
          }
      }   

    }

  });
});


// route middleware to verify a token
adminRoutes.use(function(req, res, next) {

  // check header or url parameters or post parameters for token
  var token = req.body.token || req.query.token || req.headers['x-access-token'];

  // decode token
  if (token) {
    // verifies secret and checks exp
    jwt.verify(token, app.get('superSecretAdmin'), function(err, decoded) {      
      if (err) {
        return res.status(403).json({ success: false, message: 'Failed to authenticate token.' });    
      } else {
        // if everything is good, save to request for use in other routes
        req.decoded = decoded;    
        next();
      }
    });

  } else {

    // if there is no token
    // return an error
    return res.status(403).send({ 
        success: false, 
        message: 'No token provided.' 
    });
    
  }
});

adminRoutes.get('/', function(req, res) {
    res.send('Hello! The API is admin only');
});

adminRoutes.get('/users', function(req, res) {

  UserMongo.find({  }, function (err, docs) {
    // docs is an array
    if (err) throw err;

    if (!docs) {
      res.status(204);
    } else if (docs) {
      res.json(docs);
    }   
  });

});


// show specific user
adminRoutes.get('/user/:username', function(req, res) {
  UserMongo.find({ 
    name : req.params.username
   }, function (err, user) {
    // docs is an array
    if (err) throw err;

    if (!user) {
      res.status(204);
    } else if (user) {
      res.json(user);
    }   
  });
});


// delete user
adminRoutes.delete('/user/:username', function(req, res) {
  // validate input format
 // validatorTool.validateDeleteUser(app, req, res);
  UserMongo.remove({
    name : req.params.username
  }, function(err){

    if (err) throw err;
  });
  
  res.status(204).send();

});

app.use('/admin', adminRoutes);
};